packages/node_modules/@cerebral/inferno/index.d.ts

```diff
@@ -1,12 +1,31 @@
-import { InfernoChildren } from 'inferno';
+import { InfernoChildren, Props as InfernoProps } from 'inferno';
+import InfernoComponent from 'inferno-component';
+import { ResolveValue } from 'function-tree'
+import { ControllerClass, SignalChain } from 'cerebral';
 
-export function StateContainer(opts: any): any
+type ComponentState = {} | void;
 
-export interface ContainerInterface {
-    getChildContext(): any
-    render(): InfernoChildren
+interface InfernoComponentClass<P> {
+    new (props?: P, context?: any): InfernoComponent<P, ComponentState>;
 }
-export function Container(opts: any): any
 
-export function connect(props: any, comp: (props: any) => void): any
-export function connect(props: any, comp: any): any
+interface FunctionalInfernoComponent<P> {
+    (props: P & { children?: InfernoChildren }, context?: any): InfernoChildren;
+}
+
+type InfernoCompnentType<P> = InfernoComponentClass<P> | FunctionalInfernoComponent<P>;
+
+// TODO: expose SignalsMap from cerebral
+interface SignalsMap {
+    [signalName: string]: SignalChain
+}
+
+export const StateContainer: InfernoComponentClass<{state: any, signals: SignalsMap, children?: InfernoChildren}>
+export const Container: InfernoComponentClass<{controller: ControllerClass, children?: InfernoChildren}>
+
+// TODO: expose PropsMap from cerebral
+interface PropsMap {
+    [prop: string]: ResolveValue<any>
+}
+
+export function connect<P extends PropsMap>(props: P, comp: InfernoCompnentType<Record<keyof P, any>>): InfernoComponentClass<P>
```

Comment:
This does not work because passed props are not taken into account. I know it is tempting to use the props map as source for types but until we have a solution (like [this](https://gist.github.com/gaspard/773647c5438bce74b166d4f0708a933c) one) to have proper types, having `any` does more harm then good.

Better be explicit and let users put any in there if they want to:

```ts
// Props are signal/state props, EProps are external props
export function connect<Props, EProps = {}>(propsMap: PropsMap, comp: Component<Props & EProps>): ComponentClass<EProps>
```

Also having the same type signature between `react` and `inferno` makes it easier to switch from one to the other.

Reply:
Good to know that there are `External Props` for `connect` in this case.
I think generic type `P` will be inferred from `props` argument, which will be used to enforce/type `comp`. 
The problem here is that all `Tag` functions in `props` object is typed as `(path: TemplateStringsArray | string[], ...values: any[]): Tag<T>`, which `Tag<T>` is Object rather than function here.
Then in test example:
```
  ({foo, click}) => {
    return <div onClick={click}>{foo}</div>
  }
```

`click` type will be `Tag<T>` which is not assignable to `EventHandler`... that's reason, a mapped type `Record<keyof P, any>` is in use.
@gaspard I might misunderstand `Tag<T>` here?